src/app/InteractionModelEngine.cpp

```diff
@@ -764,34 +860,131 @@ CHIP_ERROR InteractionModelEngine::CanEstablishReadTransaction(const ReadHandler
+    // Resources exhausted, since there is already some read requests ongoing on this fabric, please retry later.
+    if (usedAttributePathsInFabric + aRequestedAttributePathCount >
+            kMinSupportedPathsPerReadRequest * kMinSupportedReadRequestsPerFabric ||
+        usedEventPathsInFabric + aRequestedEventPathCount > kMinSupportedPathsPerReadRequest * kMinSupportedReadRequestsPerFabric)
     {
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(Busy));
+        return Protocols::InteractionModel::Status::PathsExhausted;
     }
 
-    return CHIP_NO_ERROR;
+    if (usedReadHandlersInFabric >= kMinSupportedReadRequestsPerFabric)
+    {
+        return Protocols::InteractionModel::Status::ResourceExhausted;
+    }
+
+    const auto evictAndUpdateResourceUsage = [&](FabricIndex fabricIndex) {
+        bool ret = TrimFabricForRead(fabricIndex);
+        countResourceUsage();
+        return ret;
+    };
+
+    //
+    // At this point, we have an inbound request that respects minimas but we still don't have enough resources to handle it. Which
+    // means that we definitely have handlers on existing fabrics that are over limits and need to evict at least one of them to
+    // make space.
+    //
+    // There might be cases that one fabric has lots of subscriptions with one interested path, while the other fabrics are not
+    // using excess resources. So we need to do this multiple times until we have enough room or no fabrics are using excess
+    // resources.
+    //
+    bool didEvictHandler = true;
+    while (didEvictHandler)
+    {
+        didEvictHandler = false;
+        didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(kUndefinedFabricIndex);
+        if (didEvictHandler)
+        {
+            continue;
+        }
+        for (const auto & fabric : *mpFabricTable)
+        {
+            // The resources are enough to serve this request, do not evict anything.
+            if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+                usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+            {
+                break;
+            }
+            didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(fabric.GetFabricIndex());
+        }
+    }
+
+    // The read requests that are not using exceeded resources will be first-come-first-served, since the read transactions will be
+    // relative short, the client is expected to retry later if the server is busy now.
+    if (usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+        usedEventPaths + aRequestedEventPathCount > eventPathCap)
+    {
+        return Protocols::InteractionModel::Status::PathsExhausted;
+    }
+    else if (usedReadHandlers >= readHandlerCap)
+    {
+        return Protocols::InteractionModel::Status::ResourceExhausted;
```

Comment:
Per spec this only gets returned for unrecoverable resource exhaustion.  This is not the right return value.